extensions/ext/GLSL_EXT_spirv_intrinsics.txt

```diff
@@ -0,0 +1,1020 @@
+        An *OpExtension* instruction will be inserted into the generated
+        SPIR-V code for each declared extension with the string parameter
+        as its `Name` operand, if it is not already present.
+
+        The `capabilities` qualifier declares a list of SPIR-V capabilities
+        that are required to use this instruction.
+        An *OpCapability* instruction will be inserted into the generated
+        SPIR-V code for each declared capability with the integer parameter
+        as its `Capability` operand, if it is not already present.
+
+    Add a new section 3.9, SPIR-V Execution Mode Declarations
+
+        The `spirv_execution_mode` and `spirv_execution_mode_id` qualifiers
+        can be used to define specific execution modes used by the
+        generated SPIR-V.
+        These qualifiers must be declared at global scope.
+
+        spirv_execution_mode-qualifier:
+        
+            `spirv_execution_mode` ( integer-constant-expression )
+            `spirv_execution_mode` ( integer-constant-expression, spirv_execution_mode-parameter-list )
+            `spirv_execution_mode` ( spirv-requirements-list, integer-constant-expression )
+            `spirv_execution_mode` ( spirv-requirements-list, integer-constant-expression, spirv_execution_mode-parameter-list )
+            
+            `spirv_execution_mode_id` ( integer-constant-expression )
+            `spirv_execution_mode_id` ( integer-constant-expression, spirv_execution_mode_id-parameter-list )
+            `spirv_execution_mode_id` ( spirv-requirements-list, integer-constant-expression )
+            `spirv_execution_mode_id` ( spirv-requirements-list, integer-constant-expression, spirv_execution_mode_id-parameter-list )
+
+        spirv_execution_mode-parameter-list:
+
+            spirv_execution_mode-parameter
+            spirv_execution_mode-parameter , spirv_execution_mode-parameter-list
+
+        spirv_execution_mode-parameter:
+
+            floating-constant
+            integer-constant
+            bool-constant
+            literal-string
+
+        spirv_execution_mode_id-parameter-list:
+
+            constant-expression
+            constant-expression , spirv_execution_mode_id-parameter-list
+
+        The `spirv_execution_mode` qualifier declares a SPIR-V execution mode
+        used by this shader.
+        An *OpExecutionMode* instruction will be inserted into the generated
+        SPIR-V code taking the first integer parameter as its `Mode` operand,
+        and subsequent operands used as its `Extra Operands`, in the same
+        order they are declared.
+
+        The `spirv_execution_mode_id` qualifier is similar to
+        `spirv_execution_mode`, but allows the use of constant expressions
+        rather than only literals, and generates an *OpExecutionModeId*
+        instruction instead.
+        Parameters to this qualifier must be constant expressions as defined
+        in section 4.3.3, Constant Expressions. 
```

Comment:
>These instructions (ExecutionMode and ExecutionModeId) aren't interchangeable in SPIR-V, you have to generate the one that corresponds to the execution mode that you're using. With that in mind, this description seems either confusing or wrong. It implies that if you want to use constant expressions then you can just switch the ID variant and carry on, but that won't work.

I mean from a glsl perspective you _can_ just switch the ID variant and carry on; whether that generates legal SPIR-V is kind of a different matter, which is why it calls out which instruction is generated in each case. I'm not sure how to make this clearer, but if you have a suggestion then that's fine...

> I guess this is to try to prevent people from declaring a specialization constant and then using it (?) but mostly it sounds like it's talking about "normal" constant expressions, like 2 + 2. We should probably copy whatever the spec currently has for layout qualifiers, which I think means accepting constant expression for all of these. I'll check what the spec currently says to prevent trying to use specialization constants in places where they can't be translated to execution modes properly.

Yes it's mostly for specialization constants; we could probably relax this, the way specialization constants have been put into the spec is a little weird and there's not really an easy way to refer to them. If you have any suggestions then I'm happy to take them.

There is one other weirdness is that glsl only talks about evaluation of _integer_ constants explicitly; there's some ambiguity about what would happen if you did e.g. 3.0f + 3.0f as a parameter. Maybe it's clarified somewhere but I couldn't see it. Since theoretically a floating point (or other type) parameter could be used here I was erring on the side of caution.

Maybe if we could just say "must not be a specialization constant" for the literal variant then this would be easier.

Reply:
Ok so I looked at this a bit more and I _think_ we could say "constant expression that doesn't involve a specialization constant" more or less. Any thoughts on that?

Reply:
> This kind of thing doesn't really make for a good language experience, and I don't think it's needed in this case. Having it say "constant expression that doesn't involve a specialization constant" is probably fine. We already have that concept, although it may not be named in the public spec yet.

Sure, I can live with that.

> What's the behaviour of the _id variant if you pass it a non-specialization constant expression? Does glslang generate a specialization constant to make it work, or is it an error or something? I think that quietly generating broken SPIR-V is not really a good option.

The switch here is between a compiling down to a literal value or using a separate OpConstant instruction (or specialization constant, or some other constant instruction). glslang isn't doing anything difficult or fancy here. If the compiler sees it used in the non _id version, then it compiles down to a literal and errors if it can't. If it sees it used in the _id version then it just points at a constant instruction in SPIR-V.

Reply:
So currently the only mention that I see of execution_mode requiring a constant is when the _id variant says that they are allowed. It would be good to say either that execution_mode requires a literal constant or that it's an error if the expression can't be resolved to a constant when generating the SPIR-V, depending on which is correct. I would then rearrange the description of execution_mode_id to say that it's like execution_mode except that it generates OpExecutionModeId. Explaining any other differences following that will help make it clear that really this is intended to insert something different, it's not just a variant of the other thing.

Alternatively, having looked at the decorate/decorate_id text below we could change this to match that. Whatever you end up with it would be good to be consistent between these two places.

Reply:
Ok, I'll have a look - making it match decorate/decorate_id might be fine and is possibly what I originally intended..

Reply:
I think it's worth spelling this out again, but I think the grammar is in fact correct, and so maybe I just need to change the text to clarify, and I should make decorate work the same way.